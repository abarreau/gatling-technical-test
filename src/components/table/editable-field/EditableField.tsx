import { FC, FormEvent, ReactNode, useState } from 'react';

/**
 * When inserted inside a table, this component allows to edit the displayed value
 * @param value the value to display
 * @param onValueChange called when the input loses focus with the updated value
 */
export const EditableField: FC<{
    value: string,
    className?: string,
    onValueChange: (value: string) => void,
    children: ReactNode
}> = ({ children, value, onValueChange, className }) => {
    const [ isBeingEdited, setBeingEdited ] = useState<boolean>(false);
    const [ formValue, setFormValue ] = useState<string>(value);

    const onChange = (e: FormEvent<HTMLInputElement>): void => setFormValue(e.currentTarget.value);

    const onBlur = () => {
        setBeingEdited(false);
        onValueChange(formValue);
    };

    if(isBeingEdited) {
        return (<td className={className}>
            <input autoFocus={true} onBlur={onBlur} type="text" value={formValue} onChange={onChange}/>
        </td>);
    } else {
        return <td className={className} onClick={() => !isBeingEdited ? setBeingEdited(true) : ''}>{ children }</td>;
    }
};