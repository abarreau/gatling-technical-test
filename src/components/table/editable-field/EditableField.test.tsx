import renderer from 'react-test-renderer';
import { EditableField } from './EditableField';
import { act } from '@testing-library/react';

describe('Table', () => {
    it('should display component', () => {
        const onValueChange = jest.fn();
        const component = renderer.create(
            <EditableField value="test" onValueChange={onValueChange}>test</EditableField>
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should display input when clicking on td', () => {
        const onValueChange = jest.fn();
        const component = renderer.create(
            <EditableField value="test" onValueChange={onValueChange}>test</EditableField>
        );

        act(() => {
            component.root.findByType('td').props.onClick();
        });

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should invoke onValueChange', () => {
        const onValueChange = jest.fn();
        const component = renderer.create(
            <EditableField value="test" onValueChange={onValueChange}>test</EditableField>
        );

        act(() => {
            component.root.findByType('td').props.onClick();
        });

        act(() => {
            component.root.findByType('input').props.onBlur();
        });

        expect(onValueChange).toHaveBeenCalled();

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
