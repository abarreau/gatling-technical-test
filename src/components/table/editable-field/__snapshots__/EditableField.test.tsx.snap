// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should display component 1`] = `
<td
  onClick={[Function]}
>
  test
</td>
`;

exports[`Table should display input when clicking on td 1`] = `
<td>
  <input
    autoFocus={true}
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
    value="test"
  />
</td>
`;

exports[`Table should invoke onValueChange 1`] = `
<td
  onClick={[Function]}
>
  test
</td>
`;
